src/config.rs

```diff
@@ -85,23 +99,104 @@ impl LayeredConfigs {
 
     /// Creates new merged config.
     pub fn merge(&self) -> config::Config {
-        let config_sources = [
-            Some(&self.default),
-            Some(&self.env_base),
-            self.user.as_ref(),
-            self.repo.as_ref(),
-            Some(&self.env_overrides),
-            self.arg_overrides.as_ref(),
-        ];
-        config_sources
+        self.sources()
             .into_iter()
-            .flatten()
+            .map(|(_, config)| config)
             .fold(config::Config::builder(), |builder, source| {
                 builder.add_source(source.clone())
             })
             .build()
             .expect("loaded configs should be merged without error")
     }
+
+    pub fn sources(&self) -> Vec<(ConfigSource, &config::Config)> {
+        let config_sources = [
+            (ConfigSource::Default, Some(&self.default)),
+            (ConfigSource::Env, Some(&self.env_base)),
+            (ConfigSource::User, self.user.as_ref()),
+            (ConfigSource::Repo, self.repo.as_ref()),
+            (ConfigSource::Env, Some(&self.env_overrides)),
+            (ConfigSource::CommandArg, self.arg_overrides.as_ref()),
+        ];
+        config_sources
+            .into_iter()
+            .filter_map(|(source, config)| config.map(|c| (source, c)))
+            .collect_vec()
+    }
+
+    pub fn resolved_config_values(
+        &self,
+        name: Option<&str>,
+        include_defaults: bool,
+    ) -> Result<Vec<(ConfigSource, Option<String>, config::Value)>, ConfigError> {
+        let mut values = vec![];
+        for (source, config) in self.sources() {
+            // Skip built-ins if not included.
+            if source == ConfigSource::Default && !include_defaults {
+                continue;
+            }
+            if let Some(name) = name {
+                // Collect any match into values.
+                match config.get::<config::Value>(name) {
+                    Ok(val) => values.push((source, Some(name.to_string()), val)),
+                    Err(config::ConfigError::NotFound { .. }) => {}
+                    Err(e) => return Err(e.into()),
+                }
+            } else {
+                // Push whole config onto values.
+                values.push((source, None, config.collect()?.into()));
+            }
+        }
+
+        // Walk through configs in reverse and remove overridden values from lower-precedence configs.
+        // Note: This could almost be done by merging into one large config, except that would lose source/ordering information.
+        let mut seen_config_names = HashSet::new();
+        let mut overridden_idxs = HashSet::new();
+        for (i, (_, name, value)) in values.iter_mut().enumerate().rev() {
+            if let config::ValueKind::Table(_) = value.kind {
+                *value = LayeredConfigs::remove_overridden_config_names(
+                    &value.to_owned().into_table()?,
```

Comment:
This borrowing is weird. It doesn't need to be mutable or owned AFAIK but when I try to call `into_table` directly it says I'm moving a value I shouldn't and something about `self`.

Reply:
Resolved. Figured out how to use `ref` to get the table value from the pattern match, without all the weird borrowing gyrations.